modules/ipi-install-replacing-a-bare-metal-control-plane-node.adoc

```diff
@@ -0,0 +1,162 @@
+  name: kni1-master-3
+  namespace: openshift-machine-api
+spec:
+  providerSpec:
+	value:
+  	hostSelector: {}
+  	image:
+    	checksum: http://172.22.0.3:6180/images/rhcos-44.81.202004250133-0-openstack.x86_64.qcow2/rhcos-44.81.202004250133-0-compressed.x86_64.qcow2.md5sum
+    	url: http://172.22.0.3:6180/images/rhcos-44.81.202004250133-0-openstack.x86_64.qcow2/rhcos-44.81.202004250133-0-compressed.x86_64.qcow2
+  	userData:
+    	name: master-user-data
+EOF
+----
++
+[NOTE]
+====
+You can define the `BareMetalHost`, `Secret`, and `MachineSet` objects in a single YAML file, for example, `sample.yaml` and create the objects using the following command:
+
+[source,terminal]
+----
+$ oc create -f sample.yaml
+----
+====
++
+. The provisioning process (by using the baremetal-operator) installs RHCOS and prepares the host to be added to the cluster:
++
+[source,terminal]
+----
+$ oc get bmh -A
+----
++
+.Example output
+[source,terminal]
+----
+NAMESPACE               NAME           STATUS PROVISIONING STATUS     CONSUMER        BMC                     HARDWARE PROFILE   ONLINE   ERROR
+openshift-machine-api   kni1-master-0  OK     provisioned             kni1-worker-0   ipmi://10.19.136.22     dell-raid          true
+openshift-machine-api   master-1       OK     externally provisioned  kni1-master-1   ipmi://172.22.0.1:6231                     true
+openshift-machine-api   master-2       OK     externally provisioned  kni1-master-2   ipmi://172.22.0.1:6232                     true
+openshift-machine-api   master-3       OK     provisioning            kni1-master-3   ipmi://172.22.0.1:6233  unknown            true
+openshift-machine-api   worker-0       OK     provisioned             kni1-worker-1   ipmi://10.19.143.29     dell-raid          true
+----
++
+. The BareMetalHost is added to the cluster:
++
+[source,terminal]
+----
+$ oc get bmh -A
+----
++
+.Example output
+[source,terminal]
+----
+NAMESPACE             NAME     STATUS PROVISIONING STATUS     CONSUMER          BMC                    HARDWARE PROFILE ONLINE ERROR
+openshift-machine-api master-1 OK     externally provisioned  kni1-master-1     ipmi://172.22.0.1:6231                  true 
+openshift-machine-api master-2 OK     externally provisioned  kni1-master-2     ipmi://172.22.0.1:6232                  true
+openshift-machine-api master-3 OK     provisioned             kni1-master-3     ipmi://172.22.0.1:6233 unknown          true
+openshift-machine-api worker-0 OK     provisioned             kni1-worker-0     ipmi://10.19.143.29    dell-raid        true
+----
++
+. Confirm that the node has been added to the cluster:
++
+[source,terminal]
+----
+$ oc get nodes
+----
++
+.Example output
+[source,terminal]
+----
+NAME             STATUS      ROLES    AGE   VERSION
+kni1-master-1    available	 master   141m  v1.18.2
+kni1-master-2    available	 master   141m  v1.18.2
+kni1-master-3    available	 master   4m2s  v1.18.2
+kni1-worker-0    available	 worker   87m   v1.18.2
+----
```

Comment:
Then, fix etcd.

Get the list of etcd pods:

```
$ oc get po -n openshift-etcd -l app=etcd
NAME                               READY   STATUS             RESTARTS   AGE
etcd-kni1-vmaster-1                3/3     Running            0          113m
etcd-kni1-vmaster-2                3/3     Running            0          111m
etcd-kni1-vmaster-3                2/3     CrashLoopBackOff   21         87m
```

Choose an etcd pod currently working and use it to restore the cluster (etcd-kni1-vmaster1 in this example)

Check the current etcd members, including the failed one:

```
$ oc exec -n openshift-etcd -c etcdctl etcd-kni1-vmaster-1 -- /bin/sh -c "etcdctl member list -w table"
+------------------+---------+----------------+---------------------------+---------------------------+------------+
|        ID        | STATUS  |      NAME      |        PEER ADDRS         |       CLIENT ADDRS        | IS LEARNER |
+------------------+---------+----------------+---------------------------+---------------------------+------------+
| 96614f1cfb922d6c | started | kni1-vmaster-0 | https://10.19.138.32:2380 | https://10.19.138.32:2379 |      false |
| ccf1a165dac53790 | started | kni1-vmaster-1 | https://10.19.138.33:2380 | https://10.19.138.33:2379 |      false |
| e58335b2a6977818 | started | kni1-vmaster-2 | https://10.19.138.37:2380 | https://10.19.138.37:2379 |      false |
+------------------+---------+----------------+---------------------------+---------------------------+------------+
```

Remove the failed one and verify it is not longer there:
```
$ oc exec -n openshift-etcd -c etcdctl etcd-kni1-vmaster-1 -- /bin/sh -c "etcdctl member remove 96614f1cfb922d6c"
Member 96614f1cfb922d6c removed from cluster 33a400c7a5a9c48c

$ oc exec -n openshift-etcd -c etcdctl etcd-kni1-vmaster-1 -- /bin/sh -c "etcdctl member list -w table"
+------------------+---------+----------------+---------------------------+---------------------------+------------+
|        ID        | STATUS  |      NAME      |        PEER ADDRS         |       CLIENT ADDRS        | IS LEARNER |
+------------------+---------+----------------+---------------------------+---------------------------+------------+
| ccf1a165dac53790 | started | kni1-vmaster-1 | https://10.19.138.33:2380 | https://10.19.138.33:2379 |      false |
| e58335b2a6977818 | started | kni1-vmaster-2 | https://10.19.138.37:2380 | https://10.19.138.37:2379 |      false |
+------------------+---------+----------------+---------------------------+---------------------------+------------+
```

After a while, the new node will be added to the cluster:

```
$ oc get po -n openshift-etcd -l app=etcd
NAME                               READY   STATUS              RESTARTS   AGE
etcd-kni1-vmaster-1                3/3     Running             0          29s
etcd-kni1-vmaster-2                3/3     Running             0          116m
etcd-kni1-vmaster-3                3/3     Running             22         92m

$ oc exec -n openshift-etcd -c etcdctl etcd-kni1-vmaster-1 -- /bin/sh -c "etcdctl member list -w table"
+------------------+---------+----------------+---------------------------+---------------------------+------------+
|        ID        | STATUS  |      NAME      |        PEER ADDRS         |       CLIENT ADDRS        | IS LEARNER |
+------------------+---------+----------------+---------------------------+---------------------------+------------+
| ccf1a165dac53790 | started | kni1-vmaster-1 | https://10.19.138.33:2380 | https://10.19.138.33:2379 |      false |
| e58335b2a6977818 | started | kni1-vmaster-2 | https://10.19.138.37:2380 | https://10.19.138.37:2379 |      false |
| fa5d82e7a7ab6ca8 | started | kni1-vmaster-3 | https://10.19.138.38:2380 | https://10.19.138.38:2379 |      false |
+------------------+---------+----------------+---------------------------+---------------------------+------------+
```

Reply:
@e-minguez do you have the commands and steps to choose a working etcd pod and using it to restore the cluster? It seems as though it's missing from your above write up, unless it's part of `oc exec -n openshift-etcd -c etcdctl etcd-kni1-vmaster-1 -- /bin/sh -c "etcdctl member list -w table"` and I'm just unaware? Thanks!

Reply:
The pod in 'CrashLoopBack' is the one not working, so basically you can choose any other pod. I've edited the comment with a more consistent example (I mixed the instructions from a previous environment, apologize)